### DISCUSSION TITLE
Auto-Merge fails with "undefined" artifact update problem for "package-lock.json"

### DISCUSSION THREAD

**Discussion by BnAmN** [renovatebot/renovate#22254] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

35.86.2

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### If you're self-hosting Renovate, tell us what version of the platform you run.

GitLab 15.10.0-ee

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Renovate should be able to merge or at least output a more describing error message.

In case my current suspicion is correct I would expect a way to disable Renovates tool releases check (in general or at least for `node-version`) or at least a way to provide a different URL to a mirror server.

### What you tried so far.

I'm running GitLab and Renovate in a self-hosted, air-gapped environment. Mirrors for NPM and other required package managers are available and Renovate is able to correctly detect dependency updates.  
(I'm still having a few problems with the GitLab registry and Renovate but that's not related to this issue.)

Now I tried to get auto-merge to work and noticed, that Renovate started to fail with the following error message inside the PR/MR:

![bug-report](https://github.com/renovatebot/renovate/assets/11967700/93a73977-0d9e-4bfd-8c52-c75b34baf885)

The CI/CD pipeline also fails at the external `renovate/artifacts` job:

![ci-pipeline-status](https://github.com/renovatebot/renovate/assets/11967700/11a6ed5e-2b3e-43b8-adda-d84df829022c)

Renovates output only shows:

```
 WARN: artifactErrors (repository=bnamn/web/ci-test-2, baseBranch=main, branch=renovate/stylus-0.x)
       "artifactErrors": [{"lockFile": "package-lock.json"}]
```

So I looked through the logs and found another error message that seems to be the root cause of the problem:

```
DEBUG: GET https://nodejs.org/dist/index.json = (code=ENOTFOUND, statusCode=-1 retryCount=2, duration=3) (repository=bnamn/web/ci-test-2, baseBranch=main, branch=renovate/stylus-0.x)
```

I currently don't have a `nodejs.org` mirror setup for the releases and the domain is currently unreachable for Renovate. I tried disabeling the tool lookup using the following configuration changes:

```
  packageRules: [
    {
      matchDatasources: ['node-version'],
      matchPackageNames: ['node'],
      enabled: false
    }
  ]
```

```
  packageRules: [
    {
      matchDatasources: ['node-version'],
      enabled: false
    }
  ]
```

```
  packageRules: [
    {
      matchPackageNames: ['node'],
      enabled: false
    }
  ]
```

None of the above worked.

-----

In addition I actually tried to set up a mirror for `nodejs.org` but my host rule gets ignored too:

```
  packageRules: [
    {
      matchDatasources: ['node-version'],
      registryUrls: ['https://nodejs.home.arpa/dist']
    }
  ]
```

Renovate still tries to resolve `nodejs.org`.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Generating package-lock.json for . (repository=bnamn/web/ci-test-2, baseBranch=main, branch=renovate/stylus-0.x)
DEBUG: Spawning npm install to create ./package-lock.json (repository=bnamn/web/ci-test-2, baseBranch=main, branch=renovate/stylus-0.x)
DEBUG: Updating lock file only (repository=bnamn/web/ci-test-2, baseBranch=main, branch=renovate/stylus-0.x)
DEBUG: No node constraint found - using latest (repository=bnamn/web/ci-test-2, baseBranch=main, branch=renovate/stylus-0.x)
DEBUG: Setting CONTAINERBASE_CACHE_DIR to /builds/bnamn/renovate-bot/renovate/cache/containerbase (repository=bnamn/web/ci-test-2, baseBranch=main, branch=renovate/stylus-0.x)
DEBUG: Using containerbase dynamic installs (repository=bnamn/web/ci-test-2, baseBranch=main, branch=renovate/stylus-0.x)
DEBUG: GET https://nodejs.org/dist/index.json = (code=ENOTFOUND, statusCode=-1 retryCount=2, duration=3) (repository=bnamn/web/ci-test-2, baseBranch=main, branch=renovate/stylus-0.x)
DEBUG: Datasource connection error (repository=bnamn/web/ci-test-2, baseBranch=main, branch=renovate/stylus-0.x)
       "datasource": "node-version",
       "packageName": "node",
       "url": undefined,
       "errCode": "ENOTFOUND"
DEBUG: lock file error (repository=bnamn/web/ci-test-2, baseBranch=main, branch=renovate/stylus-0.x)
       "err": {
         "message": "No tool releases found.",
         "stack": "Error: No tool releases found.\n    at resolveConstraint (/opt/buildpack/tools/renovate/35.86.2/node_modules/renovate/lib/util/exec/containerbase.ts:267:11)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at generateInstallCommands (/opt/buildpack/tools/renovate/35.86.2/node_modules/renovate/lib/util/exec/containerbase.ts:318:27)\n    at prepareRawExec (/opt/buildpack/tools/renovate/35.86.2/node_modules/renovate/lib/util/exec/index.ts:149:11)\n    at exec (/opt/buildpack/tools/renovate/35.86.2/node_modules/renovate/lib/util/exec/index.ts:176:39)\n    at Object.generateLockFile (/opt/buildpack/tools/renovate/35.86.2/node_modules/renovate/lib/modules/manager/npm/post-update/npm.ts:146:5)\n    at getAdditionalFiles (/opt/buildpack/tools/renovate/35.86.2/node_modules/renovate/lib/modules/manager/npm/post-update/index.ts:565:17)\n    at processBranch (/opt/buildpack/tools/renovate/35.86.2/node_modules/renovate/lib/workers/repository/update/branch/index.ts:422:31)\n    at writeUpdates (/opt/buildpack/tools/renovate/35.86.2/node_modules/renovate/lib/workers/repository/process/write.ts:159:17)\n    at update (/opt/buildpack/tools/renovate/35.86.2/node_modules/renovate/lib/workers/repository/process/extract-update.ts:213:11)\n    at Object.renovateRepository (/opt/buildpack/tools/renovate/35.86.2/node_modules/renovate/lib/workers/repository/index.ts:75:19)\n    at attributes.repository (/opt/buildpack/tools/renovate/35.86.2/node_modules/renovate/lib/workers/global/index.ts:184:11)\n    at start (/opt/buildpack/tools/renovate/35.86.2/node_modules/renovate/lib/workers/global/index.ts:169:7)\n    at /opt/buildpack/tools/renovate/35.86.2/node_modules/renovate/lib/renovate.ts:18:22"
       },
       "type": "npm"
DEBUG: No updated lock files in branch (repository=bnamn/web/ci-test-2, baseBranch=main, branch=renovate/stylus-0.x)

(...)

 WARN: artifactErrors (repository=bnamn/web/ci-test-2, baseBranch=main, branch=renovate/stylus-0.x)
       "artifactErrors": [{"lockFile": "package-lock.json"}]
```

</details>

**Comment by BnAmN** [comment]
**As a side note:**  
I haven't actually defined any `node` version constraint in my minimal test project so I was wondering why renovate tries to lookup a new `node` version anyway. My `package.json` looks like the following:

```
{
  "name": "@bnamn/ci-test-2",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "lint": "echo \"Linting...\" && exit 0",
    "test": "echo \"Testing...\" && exit 0",
    "test:ci": "echo \"Testing...\" && exit 0",
    "build": "echo \"Building...\" && exit 0",
    "build:ci": "echo \"Building...\" && exit 0"
  },
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "stylus": "^0.47.3"
  }
}
```

And I have no `.node-version` or any other contraint file.

**Comment by rarkins** [comment, marked as answer]
The artifacts error is likely due to the binarySource=install functionality. It tries to determine the best Node.js version to install before running npm, but if it cannot look up public websites then it will fail. You need to configure binarySource=global and just use whatever's pre-installed.

**Comment by BnAmN** [comment]
Thank you very much for the quick response!  
I feel a bit stupid that I didn't stumble over the `binarySource` when I searched the docs. I always came to the sections about the `node` and `node-version` data source.

-----

I configured `binarySource=global` but this results in a new problem:  
Now my custom `registryUrls` for `npm` are ignored.

![bug-report-2](https://github.com/renovatebot/renovate/assets/11967700/a7f30404-f304-4120-ae6b-1cb7bfba8116)

Maybe I still get the documentation wrong?

**Reply by rarkins** [reply]
binarySource=global has no effect on registryUrls

Please start a separate discussion

**Reply by BnAmN** [reply]
Okay, thank you.

I could solve the problem for now by adding `npmrc: 'registry=https://npm.home.arpa'` to Renovates config (as described in the docs).  
It feels a bit odd to define it in addition to the `registryUrls` but I understand that this gets out of the scope of this discussion.

Thank you again for clarification!

**Reply by rarkins** [reply]
`npmrc` is the recommended way for npm
